File: server/routes/iot.js
const express = require('express');
const router = express.Router();
const db = require('../database');

// POST /iot/reading
router.post('/reading', (req, res) => {
    const { farm_id, soil_moisture, water_level, temperature, humidity } = req.body;

    if (!farm_id) {
        return res.status(400).json({ error: 'farm_id is required' });
    }

    const sql = `INSERT INTO iot_readings (farm_id, soil_moisture, water_level, temperature, humidity) 
                 VALUES (?, ?, ?, ?, ?)`;

    db.run(sql, [farm_id, soil_moisture, water_level, temperature, humidity], function (err) {
        if (err) {
            console.error(err.message);
            return res.status(500).json({ error: 'Failed to insert data' });
        }
        res.json({
            message: 'Reading added successfully',
            reading_id: this.lastID
        });
    });
});

module.exports = router;

File: server_python/routes/iot.py
import logging
from fastapi import APIRouter, Response
from pydantic import BaseModel
from typing import Optional

from database import db_run

router = APIRouter()
logger = logging.getLogger("agro-backend")

class IoTReadingReq(BaseModel):
    farm_id: int
    soil_moisture: Optional[float] = None
    water_level: Optional[float] = None
    temperature: Optional[float] = None
    humidity: Optional[float] = None

@router.post("/reading")
def add_iot_reading(req: IoTReadingReq, response: Response):
    if not req.farm_id:
        response.status_code = 400
        return {"error": "farm_id is required"}
        
    sql = """
        INSERT INTO iot_readings (farm_id, soil_moisture, water_level, temperature, humidity) 
        VALUES (?, ?, ?, ?, ?)
    """
    
    try:
        res_ins = db_run(sql, [req.farm_id, req.soil_moisture, req.water_level, req.temperature, req.humidity])
        return {
            "message": "Reading added successfully",
            "reading_id": res_ins["lastID"]
        }
    except Exception as e:
        logger.error(f"Error inserting IOT data: {e}")
        response.status_code = 500
        return {"error": "Failed to insert data"}
